specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1441 @@
+PACKAGE_INFO_REFRENCE objects. Our Detour'd API variants walk this information to produce their
+answers seamlessly spanning static and dynamic package information for the process.
+
+```c++ (C++ish pseudocode)
+// An instance with no PACKAGE_INFO_REFERENCE (nullptr) is a placeholder for the static package graph
+class PackageGraphNode
+{
+...
+private:
+    PACKAGE_INFO_REFERENCE m_packageInfoReference{};
+    MDD_PACKAGE_DEPENDENCY_CONTEXT m_context{};
+    DLL_DIRECTORY_COOKIE m_addDllDirectoryCookie{};
+};
+
+static std::mutex g_lock;
+std::vector<PackageGraphNode> g_packageGraph;
+
+static LONG (WINAPI * TruGetCurrentPackageInfo)(...) = CurrentPackageInfo;
+LONG DynamicGetCurrentPackageInfo(...);
+...
+
+BOOL DllMain(...)
+{
+    if (DetourIsHelperProcess())
+    {
+        return TRUE;
+    }
+
+    if (dwReason == DLL_PROCESS_ATTACH)
+    {
+        DetourRestoreAfterWith();
+
+        DetourTransactionBegin();
+        DetourUpdateThread(GetCurrentThread());
+        DetourAttach(&TrueGetCurrentPackageInfo, DynamicGetCurrentPackageInfo);
+        ...
+        DetourTransactionCommit();
+    }
+    else if (dwReason == DLL_PROCESS_DETACH)
+    {
+        DetourTransactionBegin();
+        DetourUpdateThread(GetCurrentThread());
+        DetourDetach(&TrueGetCurrentPackageInfo, DynamicGetCurrentPackageInfo);
+        ...
+        DetourTransactionCommit();
+    }
+    return TRUE;
+}
+
+LONG DynamicGetCurrentPackageInfo(...PACKAGE_INFO* packageInfo...UINT32* count)
+{
+    *count = 0;
+
+    std::unique_lock<std::mutex> lock(g_lock);
+
+    for (node : g_packageGraph)
+    {
+        auto packageInfoReference = node.get();
+        if (node)
+        {
+            packageInfo += GetPackageInfo(packageInfoRefernce...);
+        }
+        else
+        {
+            packageInfo += TrueGetCurrentPackageInfo(...);
+        }
+        ++count
+    }
+    return ERROR_SUCCESS;
+}
+...
+
+HRESULT MddAddPackageDependency(...)
+{
+    ...
+
+    std::unique_lock<std::mutex> lock(g_lock);
+}
+```
+
+### 3.1.2. Add to the Package Graph
+
+`MddAddPackageDependency` has 2 responsibilities:
+
+1. Resolve a package dependency to a specific package
+2. Update the package graph
+
+Updating the package graph also requires the DLL Search Order be updated. See below for more details.
+
+```c++ (C++ish pseudocode)
+HRESULT MddAddPackageDependency(...)
+{
+    std::unique_lock<std::mutex> lock(g_lock);
+
+    auto packageFullName = ResolvePackageDependency(packageFamilyName, minVersion, architectureFilter, ...);
+    if (!packageFullName)
+    {
+        return ERROR;
+    }
+    return AddToPackageGraph(packageFullName, ...);
+}
+```
+
+#### 3.1.2.1. Resolve Package
+
+Resolve the package dependency to a (specific) package, per the following conditions:
+
+1. Packages in Package Family
+2. PackageType = Framework
+3. Packages registered to the user
+4. Highest version
```

Comment:
This versioning scheme seems insufficient. Clients need more fine-grained control over which version they take - to avoid breaking changes, for example. This is kind of the underpinning philosophy of Semantic Versioning, to understand and control what sort of changes you'll tolerate via version numbers.

NuGet supports [version ranges](https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges). So does [NPM](https://docs.npmjs.com/about-semantic-versioning).

Reply:
Hi @DefaultRyan. You can accomplish semantic versioning with MSIX though some of the mechanics differ.

For breaking changes, framework packages include a major version id in their package name e.g. ProjectReunion3 for major version 3. This has been used by various framework authors for years to support breaking changes with consumers retaining control over if and when they pick up a new major version.

There's additional differences in major version selection, prerelease and buildinfo. MSIX today doesn't support some of semver's syntax or semantics and thus some behaviors. See below for more details.  Which of these are problematic vs nice to have?

Dynamic Dependencies provides a runtime mechanism equivalent to the declarative `<PackageDependency Name="..." Publisher="..." MinVersion="...">` in an MSIX package's `appxmanifest.xml`. The API follows the same rules as `<PackageDepedency>` evaluation. Is that insufficient? In what way(s)?

## MSIX Package Identity compared with SemVer

MSIX package identity has a version of major.minor.build.revision
SemVer has a version of major.minor.revision[-prerelease][+build]

### Major version
Major in semver is a number _and_ it has functional meaning e.g. major=3 is a breaking change vs major=2. It's significant.

Major in MSIX is just a number. You accomplish the functional meaning by putting the major value in your package name, e.g. KittyVideos version 4.5.6 in semver can be expressed in MSIX as package name = "KittyVideos4" and version = 4.5.6.0. That makes for a different package family from KittyVideos3 and KittyVideos5. That's how framework packages have done it to date e.g. package full name `Microsoft.NET.Native.Runtime.2.2_2.2.28604.0_x64__8wekyb3d8bbwe` broken down into the 5-tuple:
- name = `Microsoft.NET.Native.Runtime.2.2`
- version = `2.2.28604.0`
- architecture = `x64`
- resource id = (empty)
- publisher id = `8wekyb3d8bbwe`
This gives consumers fine-grained control over taking a breaking change.

The one difference is MSIX doesn't currently provide a way to say "Version >=1.2.3.4 thru 3.* are all equally good to me" as you'd need to say `<PackageDependency ((Name="KittyVideos1" MinVersion="1.2.3.4") OR (Name="KittyVideos2" MinVersion="0.0.0.0") OR (Name="KittyVideos3" MinVersion="0.0.0.0"))>` (or something equivalent).

### Minor.Build.Revision vs Minor.Revision
These can be used equivalently if you set revision=0 in MSIX e.g. MSIX `1.2.3.0` and semver `1.2.3` are equivalent

### SemVer prerelease
MSIX has no exact match to semver's prerelease. MSIX's revision field is similar, but the sort order rules differ e.g. given a release and a prerelease MSIX sorted order would be [1.2.3.0, 1.2.3.1] whereas semver would be [1.2.3-1, 1.2.3]

### SemVer buildinfo
semver's buildinfo is informative metadata. MSIX package identity has no equivalent concept. You can add this sort of data into appxmanifest.xml e.g. ```
<Package>
...
    <BuildInfo>
        <Foo>bar</>
        ...
    </>
</>
```
but as that's not in a known namespace it's effectively invisible when Windows processes the package. You'd need to lookup the package's install location, load its appxmanifest.xml and parse out the <BuildInfo> metadata.

Reply:
I've run into surprising differences between nuget's support of semver and MSIX versions which are worth calling out as long as we're talking about differences:

* MSIX versions requires that each component be at most 65536, which is limiting when we sometimes like to encode the revision as YYMMDD
* Nuget supports Major.Minor.Patch.Revision[-PrereleaseTag], and _technically_ MSIX supports Major.Minor.Patch.Revision -- but don't ever set the Revision field in your MSIX because the store requires that the Revision field be 0.

Reply:
@jevansaks Good point on max values (it's actually 65535).

When MSIX was created in Win8 the common "DotQuadNumber" versioning scheme was adopted, i.e.
```
struct DotQuadNumber
{
    UINT64 major, minor, build, revision;
}
```
Also sometimes expressed as `UINT16[4]`, thus each element supports values of `0-65535`

Agreed it'd be handy to support a full date as numeric values like YYMMDD and YYMMDDHHMM. Not something supported today. An important enhancement to consider?

As for revision, Store requires *submitted packages* have revision=0. I believe that was (is?) for potential-patch-update reasons. That's a Store policy, not a technical constraint of MSIX itself (the old "Don't confuse Windows-the-OS with Windows-the-ecosystem" saw).

P.S. A minor nit, MSIX calls it Major.Minor.**Build**.Revision. See [PACKAGE_VERSION](https://docs.microsoft.com/windows/win32/api/appmodel/ns-appmodel-package_version) or [Windows.ApplicationModel.PackageVersion](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.PackageVersion?view=winrt-19041)

Reply:
> That's a Store policy, not a technical constraint of MSIX itself (the old "Don't confuse Windows-the-OS with Windows-the-ecosystem" saw).

Of course you are _techincally correct_ ([the best kind of correct](https://www.youtube.com/watch?v=hou0lU8WMgo)) but the distinction is lost on framework authors such as WinUI/Reunion since our primary vehicle for shipping these today is through the Store. If we plan to capture this information somewhere (and I hope we do) then we should call it out because anyone designing a versioning scheme around their nuget+MSIX packages will likely want some amount of consistency through them and the constraints of nuget + MSIX + store means you have to choose the lowest common denominator of all of them.

Reply:
Of course. Just noting _policy_ can change (and often do over time). It's a bit more involved when there's technical constraints e.g. Visual Studio can make Framework packages but (currently) it doesn't help you do that, as a matter of policy. No reason why that can't change (and hopefully will one day :-) Compared with, say, MSIX version fields limited to UINT16 due to a technical constraint

I'm not aware of any place today collating the deltas of MSIX and semver versioning. Yet :-)

Semver's been discussed a few times but given the impacts changing MSIX' versioning it's not something we'd want to undertake lightly. It would be really helpful to have feedback re benefits/painpoints and priorities.

This topic is distinct from Dynamic Dependencies and warrants further discussion. Let's continue the version discussion at [Enhance MSIX to support Semantic Versioning per http://semver.org #148](https://github.com/microsoft/ProjectReunion/issues/148)